modules/BuildTools.Tasks/CreateSourceLink.cs

```diff
@@ -0,0 +1,65 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System.IO;
+using Microsoft.Build.Framework;
+using Microsoft.Build.Utilities;
+
+namespace Microsoft.AspNetCore.BuildTools
+{
+#if SDK
+    public class Sdk_CreateSourceLink : Task
+#elif BuildTools
+    public class CreateSourceLink : Task
+#else
+#error This must be built either for an SDK or for BuildTools
+#endif
+    {
+        [Required]
+        public string RootDirectory { get; set; }
+
+        [Required]
+        public string OriginUrl { get; set; }
+
+        [Required]
+        public string DestinationFile { get; set; }
+
+        [Required]
+        public string Commit { get; set; }
+
+        [Output]
+        public string SourceLinkFile { get; set; }
+
+        public override bool Execute()
+        {
+            var rootDirectory = Path.Combine(Path.GetFullPath(RootDirectory), "*");
+            rootDirectory = rootDirectory.Replace("\\", "\\\\");
+
+            var codeSource = ConvertUrl();
```

Comment:
Add an error if the OriginUrl does not contain github.com. We only need to support github links for now, but it would be good to have this fail if we every decided to mirror to something else.